internal/terraform/transform_diff.go

```diff
@@ -69,7 +93,7 @@ func (t *DiffTransformer) Transform(g *Graph) error {
 			// run any condition checks associated with the object, to
 			// make sure that they still hold when considering the
 			// results of other changes.
-			update = true
+			update = t.hasConfigConditions(addr)
```

Comment:
I feel weird about making the DiffTransformer specifically aware of the preconditions in the config; this feels like some action at a distance that's likely to trip us up under future maintenance just has how not having no-op actions in the graph caused us to ship a subtly broken postconditions feature in the first place. :confounded: 

This very direct approach does seem reasonable as a small, surgical fix for backporting into the v1.3 branch, but I wonder if there's a different way we could arrange this in the v1.4 branch (soon after this is merged) so that this will hopefully be less of a maintenance hazard.

---

One starting idea, though I've not thought deeply about it yet:

* Add yet another optional interface that graph nodes can implement:

    ```hcl
    type GraphNodeDuringApply interface {
        HasApplyTimeActions() bool
    }
    ```
* In a new transformer after we've done all the ones that insert new nodes, prune from the graph any node which implements this interface and `HasApplyTimeActions()` returns `false`.
* If even inserting them in the first place is problematic then we could optionally optimize this by having `DiffTransformer` pre-check for the interface itself and not even insert the node in the first place. But my hope is that this wouldn't be necessary because we could do the pruning early enough that it's done by the time we get into all of the expensive transformers that do edge analysis.

Then this "are there any conditions?" logic could live in `GraphNodeApplyableResourceInstance` (and similar), closer to the `Execute` function that it's acting as a gate for. It would presumably return `false` unless the action is something other than `NoOp` _or_ there's at least one condition to check.

Since I'm explicitly not suggesting we do this refactoring as a blocker for this PR, happy to discuss this more in a more appropriate place after this is merged. :grinning:

Reply:
I tried a few different approaches here, and settled on this config reference only meant as a stop-gap. I think since we do store the plan check results, the results of those could be included in the "diff" that we're handling here. I was running into problems with inconsistent handling of the results however, because those have separate codepaths from normal changes, and this turned out to be a safer and more reliable failsafe for now.  Since unknown checks from planning also would indicate "HasApplyTimeActions", it may be worth seeing if we can incorporate the checks more closely with the plan object so they are always in sync.